File: 0097-interleaving-string/0097-interleaving-string.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        vector<int> arr(26,0);
        for(auto c:s1){
            arr[c-'a']++;
        }
        for(auto c:s2){
            arr[c-'a']++;
        }
        for(auto c:s3){
            arr[c-'a']--;
            if(arr[c-'a']<0){
                return false;
            }
        }
        int sm = accumulate(arr.begin(),arr.end(),0);
        return !sm?true:false;
    }
};

File: 0097-interleaving-string/0097-interleaving-string.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:


        # when i am at p1 at s1 and p3 == s3
        # then i can guess

        @cache
        def check(p1: int, p2: int) -> bool:
            p3 = p1 + p2
            if p3 == len(s3):
                return p1 == len(s1) and p2 == len(s2)
            #
            case_1 = check(p1 + 1, p2) if p1 < len(s1) and s1[p1] == s3[p3] else False
            case_2 = check(p1, p2 + 1) if p2 < len(s2) and s2[p2] == s3[p3] else False

            return case_1 or case_2

        return check(0, 0)